cmd/mtaBuild.go

```diff
@@ -0,0 +1,176 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/SAP/jenkins-library/pkg/command"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/piperutils"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"text/template"
+)
+
+const templateMtaYml =
+`_schema-version: "2.0.0"
+ID: "{{.ID}}"
+version: {{.Version}}
+
+parameters:
+  hcp-deployer-version: "1.0.0"
+
+modules:
+  - name: {{.Name}}
+    type: html5
+    path: .
+    parameters:
+       version: {{.Version}}-${timestamp}
+    build-parameters:
+      builder: grunt
+      build-result: dist`
+
+
+func mtaBuild(config mtaBuildOptions, commonPipelineEnvironment *mtaBuildCommonPipelineEnvironment) error {
+	log.Entry().Info("Launching mta build")
+	return runMtaBuild(config, commonPipelineEnvironment, &command.Command{})
+}
+
+func runMtaBuild(config mtaBuildOptions, commonPipelineEnvironment *mtaBuildCommonPipelineEnvironment,
+	s shellRunner) error {
+
+	prOut, pwOut := io.Pipe()
```

Comment:
whole stdout, stderr handling is not required.
Please see https://github.com/SAP/jenkins-library/blob/master/cmd/kubernetesDeploy.go#L64 for an example

Reply:
`stdout` / `stderr` handling in deed not required here. The example is insufficient (... putting everything into a byte buffer and work on that buffer later on) since we don't get any output as long as the command runs. For longer running commands like mta builds that is clearly a no-go. In cases where we don't need the output of the underlying process  - which is the case here - we can directly use `os.Stdout` and `os.Stderr`. --> I will change that accordingly.

Reply:
Not sure what you mean.
If there is no return value read, there is no need for the byte buffer, that´s right.
We can direct to the log writer. Then the return values appear nice and timely in the console.
You could also check the karma step for that.
Just run ˋpiper karmaExecuteTestsˋ with providing some dummy values for the default params. This is how it is handled in all steps so far (besides probably xsDeploy).
Is there a reason for deviating from other steps in this case? I personally cannot see any.

If you have doubts we could also look at it together if you like.

Reply:
Apparently we have some different points of view about how to use logging - especially when having some output from forked processes.

I guess the point is: output from a fork process != log output. Beside that we have to take cases into account where the subprocess already applies some log formatting.

The handling inside `kubernetesDeploy` might be good for that step, but it does IMO not fit as a general pattern/example.

1.) When collecting the output in a byte buffer and handle that output later we don't see that output in a timely manner. For long running processes (like mta builds) that is IMO a no-go. Also for short-runners that might be bad. What if we crash between launching the subprocess and delivering the output from a byte cache later? In case the output is not needed by the step we can directly use `os.Stderr`. In case the output  is needed by the step we should read the output in an asynchronous way and tee it. One end of the tee forward to `stderr` (see `xsDeploy`).

2.) The `stdout` output of a subprocess is not necessarily something that should be logged "per se". Example (maybe there are better examples): `curl https://example.org/myLongScript.sh -o -`. Why should we put the script to the log?

3.) Some tools provide normally already formatted log output (e.g. maven, which provides log through stdout, would be better to do this through stderr, but that is another story). In case we decide to forward the output to a logging framework with severity `[INFO]` we might get the case that an - already formatted - error from a subprocess get also labeled as `[INFO]` (something like `[INFO][ERROR] bad things has happened`. That is very confusing. Imaging someone reads the log, sees the `ERROR` from a forked process, but it is decorated again with `[INFO]`. So the troubleshooter thinks that is some resolved error, not the cause of a problem s?he is looking for. Such things happen ... . In case we integrate into some log analysis framework (e.g. splunk) we might also get in trouble when having double-decorated log messages.

xsDeploy is a special case where we have to do some log parsing (I don't like log parsing at all, but there is in this case no other way (on request I can explain more details)). So we have to capture the output here. On the other hand just capturing the output is insufficent since we don't get log output directly delivered. Hence we have to tee the stream ( 1.) capturing so the tool can access it 2.) ensure logging without delay).

Basically we have case-to-case decission here MO, taking the nature of the subprocess into account. In the case here I would like to propose not to send the output of the `mtaBuild` through the logging framework, but hand it over to `stderr` directly in order to avoid double-decorated log messages.

Reply:
Sending directly to stderr will lead to lost data in case we exchange the log target, e.g log via json to some logging system.
That is why we highly discourage logging via stdout.

For `xsDeploy` I understand that it is special.
For all other steps we chose so far the way to go through the logging framework to not loose the flexibility of logging targets.

Since mtabuild is a commonly used step and does not classify as an exception (like xsDeploy)  I do not see a reason not handle this differently from the other steps in this repo.

@SAP/jenkins-team2 Opinions?

Reply:
> For xsDeploy I understand that it is special.

The "special" part is that we have to do log parsing here. I basically dislike log file parsing, but sometimes there is no other way with reasonable efforts. Also here you could say: OK, tee it, one end of the tee belongs to you, but the other should go through the logging framework, too. So also there is no fundamental difference and technically there is no reason for not using a logging framework for the output from `stderr` from the forked process.

And with the json framework we log on level `INFO`, but the message from the forked process is still `ERROR`.  That is not better ...

I'm not an expert for logging frameworks, but I know e.g. spunk. In splunk there are  log-forwarders forwarding files to the logging system. In that case command line tools are configured to redirect `stderr` to a file and the log-forwarder monitors that log file. Not sure if it is like this for all the logging systems. But I guess that is a standard case. Can you provide more details for the use case in your mind.

IMO it is not that cool to have additional assumtions wrt logging framework behind inside the command line tool.

My main concern is the double-decorated log records. Example already mentioned above `[INFO][ERROR] something went wrong`.

We can let it approaches compete. In case we come in trouble we can adjust ...

Reply:
From my POV there is a way to address issues in case we see issues: Log formatting.

Thus I would suggest to go with the way most steps do it today.
If we see that end-users really dislike the output and it does not help on support side we can re-adjust using the formatting.
I would not jeopardize the possibility for log forwarding, hooks, ... which is built in in logrus. This was one of the reasons to pick this tooling.

Also in Jenkins we used decorations (e.g. timestamps) which came from Jenkins.
From a troubleshooting perspective  parallel runs can much easier be identified by adding info to a log message.
This was always a pain in the b*** on Jenkins since there it was always hard to get the source of the logging line.